### PR TITLE
[MRG+1] Fix canonicalize_url() on Python 3 and re-enable tests

### PR DESCRIPTION
**Pull request by redapple**
Bits of https://github.com/scrapy/scrapy/pull/1874 related to canonicalize_url()

Does NOT handle issues raised in https://github.com/scrapy/scrapy/issues/1941

### CODE REVIEW

File: scrapy/utils/url.py [hunk 1]
```diff
@@ -37,42 +40,114 @@ def url_has_any_extension(url, extensions):
...
+    # If supplied `encoding` is not compatible with all characters in `url`,
+    # fallback to UTF-8 as safety net.
+    # UTF-8 can handle all Unicode characters,
+    # so we should be covered regarding URL normalization,
+    # if not for proper URL expected by remote website.
+    try:
+        scheme, netloc, path, params, query, fragment = _safe_ParseResult(
+            parse_url(url), encoding=encoding)
+    except UnicodeError as e:
+        if encoding != 'utf8':
```
Comment (kmike) [inline scrapy/utils/url.py]: there are several other spellings for utf8: 

* utf_8;
* utf-8;
* U8;
* UTF;

these names are also case-insensitive. See https://docs.python.org/3.5/library/codecs.html#standard-encodings.

File: scrapy/utils/url.py [hunk 2]
```diff
@@ -37,42 +40,114 @@ def url_has_any_extension(url, extensions):
...
+    # every part should be safe already
     return urlunparse((scheme, netloc.lower(), path, params, query, fragment))
 
 
 def _unquotepath(path):
     for reserved in ('2f', '2F', '3f', '3F'):
         path = path.replace('%' + reserved, '%25' + reserved.upper())
-    return unquote(path)
+
+    if six.PY3:
```
Comment (kmike) [inline scrapy/utils/url.py]: it is slightly better to use six.PY2 because six.PY3 may become False in Python 4

File: scrapy/utils/url.py [hunk 3]
```diff
@@ -37,42 +40,114 @@ def url_has_any_extension(url, extensions):
...
+    # so we should be covered regarding URL normalization,
+    # if not for proper URL expected by remote website.
+    try:
+        scheme, netloc, path, params, query, fragment = _safe_ParseResult(
+            parse_url(url), encoding=encoding)
+    except UnicodeError as e:
+        if encoding != 'utf8':
+            scheme, netloc, path, params, query, fragment = _safe_ParseResult(
+                parse_url(url), encoding='utf8')
+        else:
```
Comment (kmike) [inline scrapy/utils/url.py]: When can it fail?

Comment (redapple) [inline scrapy/utils/url.py]: it fails for example for `test_canonicalize_url_unicode_query_string_wrong_encoding`
```
self = <tests.test_utils_url.CanonicalizeUrlTest testMethod=test_canonicalize_url_unicode_query_string_wrong_encoding>

    def test_canonicalize_url_unicode_query_string_wrong_encoding(self):
        # trying to encode with wrong encoding
        # fallback to UTF-8
>       self.assertEqual(canonicalize_url(u"http://www.example.com/résumé?currency=€", encoding='latin1'),
                                          "http://www.example.com/r%C3%A9sum%C3%A9?currency=%E2%82%AC")

/home/paul/src/scrapy/tests/test_utils_url.py:147: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/home/paul/src/scrapy/scrapy/utils/url.py:85: in canonicalize_url
    parse_url(url), encoding=encoding)
/home/paul/src/scrapy/scrapy/utils/url.py:54: in _safe_ParseResult
    quote(to_bytes(parts.query, encoding), _safe_chars),
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

text = 'currency=€', encoding = 'latin1', errors = 'strict'

    def to_bytes(text, encoding=None, errors='strict'):
        """Return the binary representation of `text`. If `text`
        is already a bytes object, return it as-is."""
        if isinstance(text, bytes):
            return text
        if not isinstance(text, six.string_types):
            raise TypeError('to_bytes must receive a unicode, str or bytes '
                            'object, got %s' % type(text).__name__)
        if encoding is None:
            encoding = 'utf-8'
>       return text.encode(encoding, errors)
E       UnicodeEncodeError: 'latin-1' codec can't encode character '\u20ac' in position 9: ordinal not in range(256)

/home/paul/src/scrapy/scrapy/utils/python.py:120: UnicodeEncodeError
```

File: scrapy/utils/url.py (hunk 1 above)
Comment (redapple) [inline scrapy/utils/url.py]: On second thought, the test is not really necessary, I can't think of an example of UTF8 not working, so testing if supplied encoding is not UTF8 does not make much sense, i.e., if it failed, then it must something else, so use UTF8 the 2nd time.

File: scrapy/utils/url.py [hunk 4]
```diff
@@ -7,15 +7,18 @@
 """
 import posixpath
 import re
+import six
 from six.moves.urllib.parse import (ParseResult, urlunparse, urldefrag,
                                     urlparse, parse_qsl, urlencode,
-                                    unquote)
+                                    quote, unquote)
+if six.PY3:
```
Comment (redapple) [inline scrapy/utils/url.py]: @kmike , probably `not six.PY2` here too, right?

Comment (kmike) [inline scrapy/utils/url.py]: right

**Comment by kmike** [comment]
Looks good, thanks @redapple for the hard work!

By the way, do you know if this is true? I think it can be a desirable property of canonicalize_url function.

``` 
url = canonicalize_url(url_original, encoding=X)
url2 = canonicalize_url(url, encoding=X)
# url is always the same as url2
```

File: scrapy/utils/url.py [hunk 5]
```diff
@@ -37,42 +40,112 @@ def url_has_any_extension(url, extensions):
...
-    - remove fragments (unless keep_fragments is True)
+    - remove query arguments with blank values (unless `keep_blank_values` is True)
+    - remove fragments (unless `keep_fragments` is True)
 
-    The url passed can be a str or unicode, while the url returned is always a
-    str.
+    The url passed can be bytes or unicode, while the url returned is
+    always a native str (bytes in Python 2, unicode in Python 3).
 
     For examples see the tests in tests/test_utils_url.py
```
Comment (kmike) [inline scrapy/utils/url.py]: I wonder if we should add a note about what is `canonicalize_url` function for: it is not for cleaning up URLs before sending them to a server, it is for URL comparison and duplicate detection, right?

Comment (redapple) [inline scrapy/utils/url.py]: Would not hurt to remind users of this, yes.
A bit of a shame that it's used by default in link extraction

**Comment by redapple** [comment]
@kmike , [`safe_url_string` is tested for this](https://github.com/scrapy/w3lib/blob/fddf599cda53adecc27da0b9b246186b3444be72/tests/test_url.py#L163) so worth a try with canonicalize_url too... and a test

**Comment by kmike** [comment]
@redapple do you want to make these changes in this PR, or should we just merge it as-is?

**Comment by redapple** [comment]
would be good in this one. but it'll have to be tomorrow at earliest for me.
Or someone else pushes (tries) new tests

**Pull request merged**